modules/extension/mbstyle/src/main/java/org/geotools/mbstyle/parse/MBFilter.java

```diff
@@ -460,4 +468,59 @@ public Filter filter() {
             throw new MBFormatException("Unsupported filter " + json);
         }
     }
+
+    /**
+     * Returns true if the input values are equal, false otherwise. The inputs must be numbers,
+     * strings, or booleans, and both of the same type. Examples: ["==", number, number]: boolean
+     * ["==", string, string]: boolean ["==", boolean, boolean]: boolean ["==", null, null]: boolean
+     *
+     * @return equal to expression
+     */
+    private Filter filterEqualTo(JSONArray array) {
+        if (array.size() != 3) {
+            throwUnexpectedArgumentCount("==", 2);
+        }
+        final FilterFactory2 ff = parse.getFilterFactory();
+        if (parse.isString(array, 1)) { // legacy filter syntax
+            String key = parse.get(array, 1);
+            Object value = parse.value(array, 2);
+            return ff.equal(ff.property(key), ff.literal(value), false);
+        } else {
+            Expression expression1 = parse.string(array, 1);
+            Expression expression2 = parse.string(array, 2);
+            return ff.equal(expression1, expression2, false);
+        }
+    }
+
+    /**
+     * Returns true if the input values are not equal, false otherwise. The inputs must be numbers,
+     * strings, or booleans, and both of the same type. Examples:["!=", number, number]: boolean
+     * ["!=", string, string]: boolean ["!=", boolean, boolean]: boolean ["!=", null, null]: boolean
+     *
+     * @return Not equals expression
+     */
+    private Filter filterNotEqual(JSONArray array) {
+        if (array.size() != 3) {
+            throwUnexpectedArgumentCount("!=", 2);
+        }
+        final FilterFactory2 ff = parse.getFilterFactory();
+        if (parse.isString(array, 1)) { // legacy filter syntax
+            String key = parse.get(json, 1);
+            Object value = parse.value(json, 2);
+            return ff.notEqual(ff.property(key), ff.literal(value), false);
+        } else {
+            // get the comparables
+            Expression expression1 = parse.string(json, 1);
+            Expression expression2 = parse.string(json, 2);
+            return ff.notEqual(expression1, expression2);
```

Comment:
@vickdw I was going to ask for your input here, this is my first looking into how expressions have been implemented and I wanted to take a consistent approach before changing anything substantially.

The above is based on the MBDecision.decisionNotEqual() which calls `parse.string(json,1)` as shown above. If I understand correctly the string method is willing to convert any literals (strings, numbers, booleans) to a string so we end up with a string equality check only. Rather than a numeric != numeric equality check.

Does this match with your understanding?

Reply:
Yes, that is correct.